### PR TITLE
Add a governor module to protect against late quorum 

### PR DESCRIPTION
**Pull request by Amxx**
This PR adds a new governor module that fixes an issue described by @Arachnid.

In cases where voters' participation is low, and quorum is far from being reached, voters might not be incentivized to participate for a proposal that they don't think will reach the quorum anyway. This would allow a "whale" voter to take over the vote at the end of the voting period. The whale would have enough tokens to reach the quorum and impose its result, leaving no time for voters to react.

This module ensures a minimal voting delay after quorum is reached. If quorum is reached late, then the proposal deadline will be extended to ensure this minimal post-quorum voting duration.

#### PR Checklist
- [x] Tests
- [x] Documentation
- [x] Changelog entry

### CODE REVIEW

**Comment by Amxx** [comment]
Note: if proposal details become packed as part of a breaking release (proposed in #2962), commit #ff0d213 should be reverted for gas optimization.

**Review by Arachnid — DISMISSED** [review]

File: contracts/governance/Governor.sol [hunk 1]
```diff
@@ -110,24 +110,34 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
      * @dev See {IGovernor-state}.
      */
     function state(uint256 proposalId) public view virtual override returns (ProposalState) {
-        ProposalCore memory proposal = _proposals[proposalId];
+        ProposalCore storage proposal = _proposals[proposalId];
```
Comment (Amxx) [inline contracts/governance/Governor.sol:113]: Note: this is more gas effective in the current format, where proposal is not packed and uses 3 slots. If proposal are packed in an upcoming breaking change, then this should be changed back to memory.

File: contracts/governance/extensions/GovernorExtendedVoting.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * @dev Extension of {Governor} for extending voting past quorum being reached.
+ *
+ * _Available since v4.X._
+ */
+abstract contract GovernorExtendedVoting is Governor {
+    using SafeCast for uint256;
+    using Timers for Timers.BlockNumber;
+
+    uint64 private _votingDelayExtension;
```
Comment (frangio) [inline contracts/governance/extensions/GovernorExtendedVoting.sol:18]: "Delay" is not the right term. `votingDelay` measures delay between proposal creation and start of the vote.

The voting "period" what is extended by this module.

File: contracts/governance/extensions/GovernorExtendedVoting.sol [hunk 3]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts vX.X.X (governance/extensions/GovernorExtendedVoting.sol)
```
Comment (frangio) [inline contracts/governance/extensions/GovernorExtendedVoting.sol:2]: Just so you know, this is added automatically so it's not necessary to include it manually in a PR.

File: contracts/governance/Governor.sol [hunk 4]
```diff
@@ -110,24 +110,34 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
...
-            return
-                _quorumReached(proposalId) && _voteSucceeded(proposalId)
-                    ? ProposalState.Succeeded
-                    : ProposalState.Defeated;
-        } else {
-            revert("Governor: unknown proposal id");
         }
+
+        return
+            _quorumReached(proposalId) && _voteSucceeded(proposalId) ? ProposalState.Succeeded : ProposalState.Defeated;
```
Comment (frangio) [inline contracts/governance/Governor.sol:140]: Personally, I think an `if-else` here would make the function look cleaner.

File: contracts/governance/extensions/GovernorExtendedVoting.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    function votingDelayExtension() public view virtual returns (uint64) {
+        return _votingDelayExtension;
+    }
+
+    /**
+     * @dev Update the voting delay extension duration. This operation can only be performed through a governance proposal.
+     *
+     * Emits a {VotingDelayExtensionSet} event.
+     */
+    function setVotingDelayExtension(uint64 newVotingDelayExtension) public onlyGovernance {
```
Comment (frangio) [inline contracts/governance/extensions/GovernorExtendedVoting.sol:78]: This should be virtual.

File: contracts/governance/extensions/GovernorExtendedVoting.sol [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+     */
+    function _castVote(
+        uint256 proposalId,
+        address account,
+        uint8 support,
+        string memory reason
+    ) internal virtual override returns (uint256) {
+        uint256 result = super._castVote(proposalId, account, support, reason);
+
+        Timers.BlockNumber storage extension = _extendedVoteEnd[proposalId];
```
Comment (frangio) [inline contracts/governance/extensions/GovernorExtendedVoting.sol:51]: We're using three different names for essentially the same concept: extension, extended vote end, extended deadline.

I wish we could find a more unified naming scheme.

Comment (Amxx) [reply contracts/governance/extensions/GovernorExtendedVoting.sol]: I tried renaming variables and contract to make everything clearer. Let me know what you think

**Comment by byterose** [comment]
Very good work!

**Review by Lucid36 — APPROVED** [review]

**Review by Lucid36 — DISMISSED** [review]

**Review by Lucid36 — APPROVED** [review]

**Review by Lucid36 — APPROVED** [review]

**Pull request merged**